WebCryptoAPI/sign_verify/hmac.js

```diff
@@ -0,0 +1,320 @@
+                        assert_equals(err.name, "InvalidAccessError", "Should have thrown InvalidAccessError instead of '" + err.message + "'");
+                    });
+
+                    return operation;
+                }, vector.name + " signing with wrong algorithm name");
+
+            }, function(err) {
+                // We need a failed test if the importVectorKey operation fails, so
+                // we know we never tested verification.
+                promise_test(function(test) {
+                    assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
+                }, "importVectorKeys step: " + vector.name + " signing with wrong algorithm name");
+            });
+        }, function(err) {
+            promise_test(function(test) {
+                assert_unreached("Generate wrong key for test " + vector.name + " failed: '" + err.message + "'");
+            }, "generate wrong key step: " + vector.name + " signing with wrong algorithm name");
+        });
+
+        all_promises.push(promise);
+    });
+
+    // Test verification with the wrong algorithm
+    testVectors.forEach(function(vector) {
+        // Want to get the key for the wrong algorithm
+        var promise = subtle.generateKey({name: "ECDSA", namedCurve: "P-256", hash: "SHA-256"}, false, ["sign", "verify"])
+        .then(function(wrongKey) {
+            return importVectorKeys(vector, ["verify", "sign"])
+            .then(function(vector) {
+                promise_test(function(test) {
+                    var operation = subtle.verify({name: "HMAC", hash: vector.hash}, wrongKey.publicKey, vector.signature, vector.plaintext)
+                    .then(function(signature) {
+                        assert_unreached("Verifying should not have succeeded for " + vector.name);
+                    }, function(err) {
+                        assert_equals(err.name, "InvalidAccessError", "Should have thrown InvalidAccessError instead of '" + err.message + "'");
+                    });
+
+                    return operation;
+                }, vector.name + " verifying with wrong algorithm name");
+
+            }, function(err) {
+                // We need a failed test if the importVectorKey operation fails, so
+                // we know we never tested verification.
+                promise_test(function(test) {
+                    assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
+                }, "importVectorKeys step: " + vector.name + " verifying with wrong algorithm name");
+            });
+        }, function(err) {
+            promise_test(function(test) {
+                assert_unreached("Generate wrong key for test " + vector.name + " failed: '" + err.message + "'");
+            }, "generate wrong key step: " + vector.name + " verifying with wrong algorithm name");
+        });
+
+        all_promises.push(promise);
+    });
+
```

Comment:
- Use a key length that is not a multiple of 8 and set the high bits
(should be cleared on import?)
The length parameter is octet count, not bit count.

- Pass in a wrong length HMAC value for verify
Added a test for this.

On Tue, Jun 21, 2016 at 1:34 AM, Jim Schaad <notifications@github.com>
wrote:

> In WebCryptoAPI/sign_verify/hmac.js
> <https://github.com/w3c/web-platform-tests/pull/3183#discussion_r67810376>
> :
>
> > +            }, function(err) {
> > +                // We need a failed test if the importVectorKey operation fails, so
> > +                // we know we never tested verification.
> > +                promise_test(function(test) {
> > +                    assert_unreached("importVectorKeys failed for " + vector.name + ". Message: ''" + err.message + "''");
> > +                }, "importVectorKeys step: " + vector.name + " verifying with wrong algorithm name");
> > +            });
> > +        }, function(err) {
> > +            promise_test(function(test) {
> > +                assert_unreached("Generate wrong key for test " + vector.name + " failed: '" + err.message + "'");
> > +            }, "generate wrong key step: " + vector.name + " verifying with wrong algorithm name");
> > +        });
> > +
> > +        all_promises.push(promise);
> > +    });
> > +
>
> Other test cases:
> Use a key length that is not a multiple of 8 and set the high bits (should
> be cleared on import?)
> Pass in a wrong length HMAC value for verify
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/w3c/web-platform-tests/pull/3183/files/b16dcea8a391dfaa70a98ed41139c24f2c9a168d#r67810376>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAE_NwVynoVygxQn-8N22aTl7e89IGsjks5qN3fvgaJpZM4I2ltT>
> .
>